### ISSUE TITLE
Cloud-init 18.4 is needed on CentOS 7 for templating used by CAPA and CAPV

### ISSUE THREAD

**Issue by akutz** [kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm#41]
/kind feature

**Describe the solution you'd like**
@detiber informed me that in v1alpha2 the fork (`18.3-52-gc5f78957-1~bddeb~18.04.1`) of cloud-init will no longer be used by the bootstrap controller. Instead the decision was made to use the `runcmd`. I would be thrilled as I originally questioned the value of the `kubeadm` module over the `runcmd`, except I've come to appreciate Jason's fork since it is **required** on CentOS 7. Cloud-init on CentOS 7 is slow to update to newer versions, and the most [recent version](http://mirror.centos.org/centos/7.6.1810//os/x86_64/Packages/) on CentOS 7.6.1810 is cloud-init 18.2.

Since cloud-init 18.4+ [is required](https://cloudinit.readthedocs.io/en/latest/topics/instancedata.html#using-instance-data) for JINJA templating, used by both CAPA and CAPV and presumably the new bootstrap controller, CentOS 7 will be incompatible with v1alpha2.

And before someone says "But Jason's fork is based on cloud-init 18.3, and it works with templating," yes, yes it does. But that's the result of a terribly happy accident. Jason just so happened to have forked cloud-init off of `master` between 18.3 and 18.4 **after** templating was merged into the `master` branch.

The sites [rpm2find](https://rpmfind.net/linux/rpm2html/search.php?query=cloud-init) and [pkgs.org](https://pkgs.org/download/cloud-init) have lists of the RPMs for cloud-init, and it's unclear if any of the newer packages are compatible with CentOS 7.

For machine images based on CentOS 7 to work with v1alpha2, someone needs to either fork cloud-init 18.4+ (ideally 19.1, the latest release) for CentOS 7, or document how to install cloud-init 18.4+ on CentOS 7.

**Comment by akutz** [comment]
Hi @ncdc,

I filed this as a feature, but it could also be seen as a regression from v1alpha1. I can refile it if you like.

**Comment by neolit123** [comment]
while re-inventing the wheel is probably something we should not be doing, it seems to me cloud-init can  be problematic. i had concerns about it, as a bystander back when CAP* decided to go for it, but didn't have anything concrete to bring up.

are there plans to move away from cloud-init completely or search for alternatives?

**Comment by akutz** [comment]
Hi @neolit123,

I think cloud-init is fine. It's the distro that lags behind. And the fact is, any alternative is going to introduce an entirely new set of problems. Cloud-init is pervasive and works well. It's an issue of CentOS, and I'm about to post a `Dockerfile` that builds the RPM we need.

**Comment by akutz** [comment]
FWIW, I have a lot of experience with cloud-init and Ignition, and cloud-init is very mature and works very well. Anything that spans most Linux distributions is going to encounter occasional issues across versions of a given distribution.

**Comment by neolit123** [comment]
> Anything that spans most Linux distributions is going to encounter occasional issues across versions of a given distribution.

that is true. i common solution is to use the lowest version possible that conforms all distros, but if a new feature is really required a fork will be needed.

**Comment by akutz** [comment]
> if a new feature is really required a fork will be needed.

Except it's not _new_ per se as @detiber's fork in v1alpha1 already has it. This is kind of a regression IMO. Anyway, I'm producing a Dockerfile that can create the RPM for CentOS from the cloud-init repo. At least that way the provenance is ensured.

**Comment by akutz** [comment]
Hi @neolit123,

This [branch](https://github.com/akutz/cluster-api-provider-vsphere/tree/feature/cloud-init-centos-7/hack/tools/cloud-init/centos7) of CAPV includes a [`Dockerfile`](https://github.com/akutz/cluster-api-provider-vsphere/blob/feature/cloud-init-centos-7/hack/tools/cloud-init/centos7/Dockerfile) that can be used to build cloud-init for CentOS 7:

```shell
$ docker run -it --rm \
  -v "$(pwd):/build" \
  gcr.io/cluster-api-provider-vsphere/extra/cloud-init/centos \
  /bin/bash -c 'make rpm && cp *.rpm /build'
python2 ./packages/brpm --distro=redhat
Archived the code in u'/tmp/rpmbuilddlXYQk/rpmbuild/SOURCES/cloud-init-19.1-24-g07b17236.tar.gz'
Created spec file at '/tmp/rpmbuilddlXYQk/rpmbuild/SPECS/cloud-init.spec'
Running 'rpmbuild' in '/tmp/rpmbuilddlXYQk/rpmbuild'
Wrote out redhat package '/cloud-init/cloud-init-19.1+24.g07b17236-1.el7.noarch.rpm'
Wrote out redhat package '/cloud-init/cloud-init-19.1+24.g07b17236-1.el7.src.rpm'
```

That will create two RPMs for cloud-init in the current, local directory:

```shell
$ ls *.rpm
-rw-r--r--  1 akutz  staff   881K Jul 17 18:51 cloud-init-19.1+24.g07b17236-1.el7.noarch.rpm
-rw-r--r--  1 akutz  staff   1.0M Jul 17 18:51 cloud-init-19.1+24.g07b17236-1.el7.src.rpm
```

**Comment by akutz** [comment]
Please bear in mind that I haven't had time to test this yet. I'll do that later tonight or tomorrow morning. It's my 15th anniversary today, and I'm told I'm taking her to dinner. I mean, I took her to the beach last weekend; I have to do dinner too!? :)

But seriously, I will test this in the morning if someone else doesn't before then. I'm logging off for the night. See y'all tomorrow!

**Comment by moshloop** [comment]
I think there are multiple problems with this:

* Using a forked/self-built package as critical as cloud-init has implications for the security and supportability of the resulting images. I am pretty sure doing this will invalidate any support contracts with the OS vendor e.g. Canonical, RedHat, Amazon (for Amazon Linux), By forking the package we also to take responsibility for fixing CVE's within 24 - 48 hours of release.

* Using the least common set of features has implications for the different operating systems  and versions this provider can support, especially if there are no E2E tests - Ignition based OS's (while unsupported today, have a fairly simple path to adding support if the least common denomination is used)

Templating can be handled in multiple different ways without implications on supportability and security:
* bash script
* go helper app 
* adding templating supporting directly into kubeadm

**Comment by akutz** [comment]
I’m not building from a fork. It’s built using the cloud init process. They just don’t provide a package.

Additionally, cloud init is mature, and providing an alternative is not as easy as you make it sound. As you said, this is a critical piece of the bootstrap process and should not be removed just because it needs to be built.

**Comment by michaelgugino** [comment]
My expectation is we support the major distributions without having to override system packages.  It would be most helpful to have a clear proposal outlining the things we need cloud-init to do, and then evaluate what might be better alternatives (whether using features supported by all cloud-inits, or having cloud-init just execute some helper script/bin).

**Comment by moshloop** [comment]
As far as I know, each distribution creates their own "fork" of cloud-init with config specific to them, so by building from upstream directly, we are deviating from the norm.

CVE's that are under security embargo would also first be patched first into the forks, and only once the embargo is lifted patched upstream.

Instead of cloud-init generating data in the pre-network phase (so that it can be used to configure the network), interfaces can be brought up post the cloud-init networking phase in runcmd.

**Comment by akutz** [comment]
> Instead of cloud-init generating data in the pre-network phase (so that it can be used to configure the network), interfaces can be brought up post the cloud-init networking phase in runcmd.

No it cannot. Has to happen in the local stage. I can host a call to describe the workflow for anyone interested. 

Additionally, let’s please focus this issue on cloud init 18.4+ for CentOS. Discussing alternate bootstrap methods should be its own issue.

**Comment by moshloop** [comment]
> Has to happen in the local stage 

Why ? I know it is definitely possible as that was the approach I used to bring up multiple interfaces with different VLAN's / MTU's which netplan 2 doesn't handle well.

**Comment by akutz** [comment]
>  I know it is definitely possible

Except so much of the rest of cloud-init's features break as cloud-init asserts the network should be online by the network/init stage (post-local). This includes templating.

Anyway, I've pushed images for:

* `gcr.io/cluster-api-provider-vsphere/extra/cloud-init/centos:7.3.1611`
* `gcr.io/cluster-api-provider-vsphere/extra/cloud-init/centos:7.4.1708`
* `gcr.io/cluster-api-provider-vsphere/extra/cloud-init/centos:7.5.1804`
* `gcr.io/cluster-api-provider-vsphere/extra/cloud-init/centos:7.6.1810`
* `gcr.io/cluster-api-provider-vsphere/extra/cloud-init/centos:latest`

All images have cloud-init release 19.1 RPMs in the working directory. Any of them can be obtained via the following command:

```shell
$ docker run --rm -v "$(pwd):/build" \
  gcr.io/cluster-api-provider-vsphere/extra/cloud-init/centos \
  sh -c 'cp *.rpm /build'

$ ls *.rpm
-rw-r--r--  1 akutz  staff   874K Jul 18 13:38 cloud-init-19.1-1.el7.noarch.rpm
-rw-r--r--  1 akutz  staff   999K Jul 18 13:38 cloud-init-19.1-1.el7.src.rpm
```

The [branch](https://github.com/akutz/cluster-api-provider-vsphere/blob/feature/cloud-init-centos-7/hack/tools/cloud-init/centos7) has been updated to support building against different versions of CentOS.

**Comment by ncdc** [comment]
I filed https://bugzilla.redhat.com/show_bug.cgi?id=1733346 to see if we can get this either backported or the version upgraded in RHEL/Centos

**Comment by moshloop** [comment]
Standardising on minimal cloud-init + something like https://github.com/coreos/afterburn might be a good middle ground that supports the widest variety of OS's. 

Afterburn doesn't support the vsphere metadata, but it should be straightforward to add it.

Distributing afterburn as a static binary is straightforward and getting packages built and maintained upstream, in theory, should be lower friction than a cloud-init upgrade.

For a golang option there is https://github.com/banzaicloud/satellite but it would also need a bit of work to return all meta-data and work for vsphere.

If this is something that has rough consensus, I am prepared todo the actual work (both short-term in terms of code, and long-term in getting it upstreamed as native packages)

**Comment by moshloop** [comment]
@LaneWolf could you comment on cloud-init 18.5 support for RHEL 7 & 8 ?

**Comment by ncdc** [comment]
I'd like to wait a bit and see if we get any comments on the bz I opened before we decide to search for alternatives.

**Comment by ncdc** [comment]
RHEL 7.7 and 8.1, and presumably the corresponding CentOS versions when they are available, have cloud-init 18.5.

**Comment by detiber** [comment]
Closing since this is now resolved with the latest images built in the image-builder
/close
